### ISSUE TITLE
Provide system property gwt.codeserver.host for server side GWT-RPC

### ISSUE THREAD

**Issue by jnehlmeier** [gwtproject/gwt#9857]
**GWT version:** HEAD

---

##### Description

Currently you can start server side GWT-RPC with system property `gwt.codeserver.port` to activate downloading serialization policy files from code server at `http://localhost:<port>`.

You can override the corresponding method to change the host but then you loose access to configured `<port>` and have to write code to read and validate that system property again.

For consistency GWT should also provide `gwt.codeserver.host` with following behavior:
- if only `gwt.codeserver.port` is provided, `http://localhost:<port>` is used
- if only `gwt.codeserver.host` is provided, `<host>:9876` is used
- if both are provided, `<host>:<port>` is used
- if `gwt.codeserver.host` does not contain a protocol, `http://` is assumed. Otherwise protocol is checked to be `http://` or `https://`
- if no property is provided, downloading of policy file is inactive

My personal use case is running Jetty within a docker container while the code server runs directly on the host. In that case you would use `http://docker.host.internal` as `gwt.codeserver.host`.

**Comment by tbroyer** [comment]
I'd rather have a full origin (scheme, host, port) than a _host_ that can optionally include the scheme but never the port.

In that case, either
* you provide `gwt.codeserver.host` (or `.origin`, or possibly even a `.baseurl`) and then the `gwt.codeserver.port` is entirely ignored (could emit a warning if both are provided); or
* you provide `gwt.codeserver.port`, which is equivalent to providing `gwt.codeserver.host=http://localhost:${port}`

I wonder if this wouldn't open a security hole though: couldn't the policy be used to somehow "redirect" some calls to other classes and create RCEs? I think this is the reason `http://localhost:` is hard-coded.

**Comment by jnehlmeier** [comment]
Yeah I would also be fine with one property providing scheme, host and port and possibly mark `gwt.codeserver.port` deprecated. But given that we already have a port property I just wanted to augment it with an additional one. Maybe `gwt.codeserver.host` isn't the perfect name.

> I wonder if this wouldn't open a security hole though: couldn't the policy be used to somehow "redirect" some calls to other classes and create RCEs? 

I don't think it would be less secure assuming the property is not configured in production. For this to be a security hole you would need to kill the application and restart it with added system property or you need to figure out a way to set the system property at runtime from within the running VM by injecting and executing Java code. 

- If you can inject and execute code then you don't need to change GWTs system properties anymore.
 
- If you can restart the application with added system properties I would argue that you likely have enough permissions to also change the *.war file, replace the hardcoded string pointing to localhost and start that modified *.war file

> I think this is the reason http://localhost: is hard-coded.

You can still override the method in your own code and do whatever you want. While I believe it should not be necessary (see above) we could add a constructor to `RemoteServiceServlet` taking a whitelist of allowed scheme/host/port combinations (possibly a regex). Then you could dynamically configure the server using the system properties within the bounds of that whitelist.

**Comment by tbroyer** [comment]
It's a matter of _security by default_. `localhost` is assumed to be safe, anything else could give you bad serialization policies, or possibly the trafic be intercepted to inject those (if not protected by HTTPS).

**Iff** the serialization policy permits "rerouting" RPC calls to arbitrary classes (when deobfuscating the payload), it by itself creates an RCE, so it's not safe giving control of the serialization policy to anyone else. If you can do it by overriding a method, great, you take responsibility. If it's as easy as passing a system property, it's another matter. Sure you're supposed to give it a "safe" value, but mistakes happen, and more importantly users wouldn't be sufficiently aware of the risks.

This is all conditioned to what using arbitrary serialization policies allows, so **I might be wrong.**

Having another boolean property to switch from `localhost` to `docker.host.internal` could solve your problem while keeping things safe.

**Comment by jnehlmeier** [comment]
> This is all conditioned to what using arbitrary serialization policies allows, so I might be wrong.

You can alter the list of classes that GWT-RPC should serialize/deserialize including fields in these classes. So you would need to first make an evil class somehow discoverable by classloader, then inject an evil policy file before the first GWT-RPC request is made (since the policy is cached) and finally do a GWT-RPC request with a payload that tells the server to deserialize that evil class.

If today an application is accidentally deployed with `gwt.codeserver.port` being set (as you assume misconfiguration is easy), then two things can happen:
- anyone having access to the host running the server (if it is not running in its own namespace) can launch a process listening on that port and provide a policy file
- if no process listens on that port, GWT-RPC falls back to a legacy policy allowing all classes implementing IsSerializable. So given the description above, an evil class could simply implement that interface. But if your own app only uses Serializable then it won't work anymore and you would quickly fix the misconfiguration. But harm might already be done.

IMHO the fact that we allow such a trigger to make development easier isn't really secure by default. Even with `localhost` if you assume misconfiguration. 

I don't see a different situation if GWT would also have a `host` property. If the host is misconfigured you end up with the legacy policy described above (as the url causes a timeout or provides anything but not a parseable policy file) and your app won't work. If no property is configured then configuring it to a specific value by an attacker requires the same work as with the `port` property today.

**Comment by tbroyer** [comment]
To make things clear: I was not talking about production environments (an RCE on a dev machine can do harm too), and not necessarily an "evil class", but more likely an "unsafe" class, and possibly taking advantage of a vulnerability on that class. This would be very much a _fishing_ situation where whether "bad things" happen highly depend on the payload of the requests, that you don't control as an attacker. But just routing values to some DNS resolving constructor could allow one to gather information (can't remember if RPC will only ever call zero-argument constructors or not).

Anyway, feel free to dismiss that claim if this seems impossible or way too improbable; or possibly just document it if that seems reasonable enough as a protection measure.
